source/common/http/conn_manager_utility.cc

```diff
@@ -92,8 +92,13 @@ ConnectionManagerUtility::MutateRequestHeadersResult ConnectionManagerUtility::m
   if (!Utility::isUpgrade(request_headers)) {
     request_headers.removeConnection();
     request_headers.removeUpgrade();
+
     if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.sanitize_te")) {
-      request_headers.removeTE();
+      if (Grpc::Common::isGrpcRequestHeaders(request_headers)) {
```

Comment:
I think TE is a hop by hop header, and if we want to send it upstream we should configure it in request_headers_to_add.  For example if someone has a web gRPC filter and is downgrading from HTTP/2 gRPC to HTTP/1.1 they wouldn't necessarily support trailers on the upstream to envoy hop.
cc @RyanTheOptimist for thoughts
/wait

Reply:
Yeah, I agree with @alyssawilk. It seems like request_headers_to_add would be a more standards complaint solution.

Reply:
After catching up on #30535, I think I've changed my mind. I think that Envoy should send "TE: trailers" to the upstream when the downstream client sent it to Envoy. So instead of adding adding "trailers" when it's a gRPC request, I wonder if we should change the TE sanitization logic to preserve trailers, if present?

Reply:
@alyssawilk we could remove the TE header when making an HTTP/1 request and keep it as `trailers` when doing an HTTP/2 or 3 request.

@RyanTheOptimist I like this idea but since only `trailers` can be present in HTTP/2 and 3, why not always setting it to this value?

Reply:
It sounds like we're converging, which is great. it looks like the current patch set preserves trailers, if it was present. Since the TE header can have multiple values, do we need to check that trailers is one of the values in the list as opposed to just checking the string value of the field?

Reply:
You're right! It's updated.